Realm/RLMRealm.mm

```diff
@@ -360,123 +313,186 @@ static id RLMAutorelease(id value) {
+                RLMSetErrorOrThrow(RLMMakeError(RLMErrorFileExists, ex), outError);
+                break;
+            case RealmFileException::Kind::AccessError:
+                RLMSetErrorOrThrow(RLMMakeError(RLMErrorFileAccessError, ex), outError);
+                break;
+            default:
+                RLMSetErrorOrThrow(RLMMakeError(RLMErrorFail, ex), outError);
+                break;
+        }
+    }
+    catch (std::system_error const& ex) {
+        RLMSetErrorOrThrow([NSError errorWithDomain:NSPOSIXErrorDomain code:ex.code().value() userInfo:nil], outError);
     }
+    catch (const std::exception &exp) {
+        RLMSetErrorOrThrow(RLMMakeError(RLMErrorFail, exp), outError);
+    }
+    return nullptr;
+}
+
++ (instancetype)realmWithConfiguration:(RLMRealmConfiguration *)configuration error:(NSError **)error {
+    configuration = [configuration copy];
+    Realm::Config& config = configuration.config;
+
+    NSString *path = configuration.path;
+    bool dynamic = configuration.dynamic;
+    bool readOnly = configuration.readOnly;
 
     // try to reuse existing realm first
-    RLMRealm *realm = RLMGetThreadLocalCachedRealmForPath(path);
-    if (realm) {
-        if (realm->_readOnly != readOnly) {
-            @throw RLMException(@"Realm at path already opened with different read permissions", @{@"path":realm.path});
-        }
-        if (realm->_inMemory != inMemory) {
-            @throw RLMException(@"Realm at path already opened with different inMemory settings", @{@"path":realm.path});
-        }
-        if (realm->_dynamic != dynamic) {
-            @throw RLMException(@"Realm at path already opened with different dynamic settings", @{@"path":realm.path});
+    if (config.cache || dynamic) {
+        RLMRealm *realm = RLMGetThreadLocalCachedRealmForPath(config.path);
+        if (realm) {
+            auto const& old_config = realm->_realm->config();
+            if (old_config.read_only != config.read_only) {
+                @throw RLMException(@"Realm at path already opened with different read permissions", @{@"path":realm.path});
+            }
+            if (old_config.in_memory != config.in_memory) {
+                @throw RLMException(@"Realm at path already opened with different inMemory settings", @{@"path":realm.path});
+            }
+            if (realm->_dynamic != dynamic) {
+                @throw RLMException(@"Realm at path already opened with different dynamic settings", @{@"path":realm.path});
+            }
+            return RLMAutorelease(realm);
         }
-        return RLMAutorelease(realm);
     }
 
-    NSData *key = configuration.encryptionKey ?: keyForPath(path);
-    realm = [[RLMRealm alloc] initWithPath:path key:key readOnly:readOnly inMemory:inMemory dynamic:dynamic error:error];
-    if (error && *error) {
-        return nil;
+    if (!configuration.encryptionKey) {
+        configuration.encryptionKey = keyForPath(path);
```

Comment:
No, setting it on `configuration` updates `config` and avoids duplicating the conversion logic.